src/haskell/parser/src/CharDict.hs

```diff
@@ -0,0 +1,24 @@
+module CharDict
+       ( CharDict
+       , make
+       , memoize
+       , (!)
+       )
+  where
+
+import qualified Data.Array as Array
+import           Data.Maybe (fromMaybe)
+
+newtype CharDict a = CharDict { getCharDict :: Array.Array Char a }
+
+memoize :: (Char -> a) -> CharDict a
+memoize f =
+    CharDict $ Array.array ('\0', '\255') [(c, f c) | c <- ['\0'..'\255']]
+
+make :: [(Char,a)] -> a -> CharDict a
+make dict defaultValue =
```

Comment:
There are a lot of 'make' functions that get exported. Maybe it would be a better idea to name them differently (like mkCharDict in this case) so they do not conflict when importing multiple packages.